File: python/array.py
# 1.数组的插入、删除、按照下标随机访问操作；
# 2.数组中的数据类型是Int
#
# Author:Lee

class Array():

    def __init__(self):
        '''数组类初始化方法.'''
        self.__data = []  # 数据存储List

    def find(self, index):
        '''数组的查找方法.

        参数：
            index:将要查找的数据的下标

        返回：
            如果查找成功，则返回找到的数据
            如果查找失败，则返回False
        '''
        if index > len(self.__data) or index < 0:
            return False
        else:
            return self.__data[index]

    def delete(self, index):
        '''数组的删除方法.

        参数：
            index:将要删除的数据的下标

        返回：
            如果删除成功，则返回True
            如果删除失败，则返回False
        '''
        if index > len(self.__data) or index < 0:
            return False
        else:
            self.__data.pop(index)
            return True

    def insert(self, index, value):
        '''数组插入数据操作.

        参数：
            index:将要插入的下标
            value：将要插入的数据

        返回：
            如果插入成功，则返回True
            如果插入失败，则返回False
        '''
        if index > len(self.__data) or index < 0:
            return False
        else:
            self.__data.insert(index, value)
            return True

    def insertToTail(self, value):
        '''直接在数组尾部插入数据.

        参数：
            value:将要插入的数据
        '''
        self.__data.append(value)

    def printAll(self):
        '''打印当前数组所有数据'''
        print(self.__data)

File: csharp/05-array/Array.cs
using System;

namespace algo05_array
{
    public sealed class Array<T> where T : IComparable<T>
    {
        private T[] _data;
        private readonly int _capacity;
        private int _length;

        public Array(int capacity)
        {
            _data = new T[capacity];
            _capacity = capacity;
            _length = 0;
        }

        // length of list
        public int Length => _length;

        // insert a new element at specified index (index start from 0)
        public void Insert(int index, T newElem)
        {
            if (_length == _capacity)
            {
                throw new OutOfMemoryException("List has no more space");
            }

            if (index < 0 || index > _length)
            {
                throw new IndexOutOfRangeException("Index was outside the bounds of the list");
            }

            // to loop array from end until finding the target index
            for (int k = _length; k > index; k--)
            {
                _data[k] = _data[k - 1];
            }

            _data[index] = newElem;

            _length++;
        }

        // get an element base on index
        public T Find(int index)
        {
            if (index < 0 || index > _length - 1)
            {
                throw new IndexOutOfRangeException("Index was outside the bounds of the list");
            }

            return _data[index];
        }

        // search the node which matches specified value and return its index (index start from 0)
        public int IndexOf(T val)
        {
            if (_length == 0) return -1;
            if (_data[0].Equals(val)) return 0;
            if (_data[_length - 1].CompareTo(val) == 0) return _length - 1;

            int start = 1;
            while (start < _length - 1)
            {
                if (_data[start].CompareTo(val) == 0) return start;
                start++;
            }

            return -1;
        }

        // delete an node which is on the specified index
        public bool Delete(int index)
        {
            if (index < 0 || index > _length - 1)
            {
                throw new IndexOutOfRangeException("Index must be in the bound of list");
            }

            T deletedElem = _data[index];
            if (index < _length - 1)
            {
                for (int k = index; k < _length; k++)
                {
                    _data[k] = _data[k + 1];
                }
            }

            _length--;

            return true;
        }

        // delete an node 
        public bool Delete(T val)
        {
            int index;
            for (index = 0; index < Length; index++)
            {
                if (_data[index].CompareTo(val) == 0) break;
            }

            if (index >= Length) return false;

            return Delete(index);
        }

        // clear list
        public void Clear()
        {
            _data = new T[_capacity];
            _length = 0;
        }
    }
}